source/common/router/router.cc

```diff
@@ -305,8 +305,21 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
 }
 
 Http::ConnectionPool::Instance* Filter::getConnPool() {
+  // Choose protocol based on cluster configuration and downstream connection
+  // Note: Cluster may downgrade HTTP2 to HTTP1 based on runtime configuration.
+  auto features = cluster_->features();
+
+  Http::Protocol protocol = (features & Upstream::ClusterInfo::Features::HTTP2)
+                                ? Http::Protocol::Http2
+                                : Http::Protocol::Http11;
+  if (features & Upstream::ClusterInfo::Features::USE_DOWNSTREAM_PROTOCOL) {
+    const Optional<Http::Protocol>& downstream_protocol = callbacks_->requestInfo().protocol();
+    if (downstream_protocol.valid()) {
```

Comment:
IIRC protocol must be valid here. I think you can skip the if check.

Reply:
Taking your word on this :-)